design: BGP FRR integration (#832)
metallb/metallb: design/speaker.md

# Speaker Design

Metallb’s speaker is deployed as a container with a Dockerfile specified
[here](https://github.com/metallb/metallb/blob/main/speaker/Dockerfile) and
running the [“speaker” Golang application](https://github.com/metallb/metallb/tree/main/speaker).
This "speaker" application is the main application that manages advertisement of
prefixes based on the creation of kubernetes services. The docker image and the
golang executable get built through [tasks.py](https://github.com/metallb/metallb/blob/main/tasks.py)
as part of the Metallb build process.

The "speaker" application supports a Layer 2 mode and a BGP mode. The Layer 2

> mode advertises prefixes using Gratuitous ARP, whereas the BGP mode

Author: I will clarify this in the next revision.

                                                            │  <<use>>
                                                    ┌───────▼───────┐
                                                    │               │
                                                    │    session    │
                                                    │ <<interface>> │
                                                    │               │
                                                    └───────▲───────┘
                                                            │ <<implements>>
                                                            │
                                                    ┌───────┴───────┐
                                                    │               │
                                                    │     bgp       │
                                                    │  <<package>>  │
                                                    │               │
                                                    └───────────────┘

```

### main

The main package contains the main() entrypoint for the speaker application. It

> creates an L2 and/or BGP controller, a speakerList and a k8s client. The k8s

Reviewer: IMO we should use either "Layer 2" or "L2" consistently throughout the document.

Author: Changed to "Layer 2"

client watches for changes to the MetalLB config map, service configuration, and
the node configuration and calls the appropriate protocol handler (e.g. L2, BGP)
via the “Protocol” Interface after ensuring the change is valid.

The BGP implementation of the “Protocol” interface can be found
[here](https://github.com/metallb/metallb/blob/main/speaker/bgp_controller.go).
This handles changes to the config map, service, and node configuration for BGP
pools. Based on these changes, BGP sessions will be established/destroyed and
prefixes will be advertised via the bgp package.

The Layer 2 implementation of the "Protocol" interface can be found
[here](https://github.com/metallb/metallb/blob/main/speaker/layer2_controller.go).

### bgp

The bgp package manages BGP sessions and advertisements. A BGP session is
created by the New() method and prefixes are advertised by the Set() method
which is part of the session interface.

BGP sessions are uniquely identified and configured by the parameters of the
New() method which are currently:

> * **addr**: IP address and port of peer

Author: updated to add this field.

* **asn**: Local ASN
* **routerID**: Specify the BGP Router ID of the session
* **peerASN**: ASN of the router that we are connecting to
* **holdTime**: BGP Hold Time specifies how long to maintain sessions for
* **Password**: TCP MD5 sock option for connection
* **myNode**

Advertisements are made for a session via the Set() method.

### speakerlist

speakerlist uses the Hashicorp memberlist package to create a cluster of
speakers, manage membership of this cluster and to detect member failures using

> the “gossip” protocol. This is only used in the speaker's L2 mode.

Reviewer: ```suggestion
a gossip protocol. This is only used in the speaker's L2 mode.
```

"Gossip protocols" is a category of protocols rather than the name of a specific protocol.

### config

> Parses configmap into an internal representation of the configuration described

Reviewer: - Missing article before "configmap".
- Let's use either "config map" or "configmap" consistently.

Author: I added the article and changed to "ConfigMap" as this is the proper type of the object.

by structs in the config package.

### k8s

> Kubernetes client/watcher

Reviewer: ```suggestion
A Kubernetes client/watcher.
```

## BGP session creation

> New BGP sessions are established by calling newBGP() in the speaker "main"

Reviewer: ```suggestion
New BGP sessions are established by calling newBGP() in the speaker's "main"
```

> package which calls bgp.New() returning a session interface. The session

Reviewer: ```suggestion
package which calls bgp.New() which in turn returns a session interface. The session
```
